File: scripts/scrape_matthew.py
import os
import re
import requests
from bs4 import BeautifulSoup
from supabase import create_client, Client

# Supabase setup
url = os.environ.get("VITE_SUPABASE_URL", "https://avrnbefzxtznpodugacz.supabase.co")
key = os.environ.get("VITE_SUPABASE_SERVICE_ROLE_KEY") or os.environ.get("VITE_SUPABASE_ANON_KEY", "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImF2cm5iZWZ6eHR6bnBvZHVnYWN6Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3ODM2MTU3MzksImV4cCI6MjA5OTE5MTczOX0._wZp4bK1b2XGSYYUWzQTw2mkyCyKvwOi6iyIIuauRKI")
supabase: Client = create_client(url, key)

TRANSLATION_ID = 1
BOOK_ID = 'mat'

def scrape_chapter(chapter_num):
    print(f"Scraping chapter {chapter_num}...")
    if chapter_num == 1:
        url = "https://augustino.net/tin-mung-theo-thanh-mat-theu"
    else:
        url = f"https://augustino.net/tin-mung-theo-thanh-mat-theu-chuong-{chapter_num}"
        
    response = requests.get(url)
    if response.status_code != 200:
        print(f"Failed to fetch chapter {chapter_num}: HTTP {response.status_code}")
        return []

    soup = BeautifulSoup(response.text, 'html.parser')
    
    # In augustino.net, the content is inside a div with id "page-content" or class "content"
    content_div = soup.find('div', id='page-content') or soup.find('div', class_='content')
    if not content_div:
        print(f"Could not find content div for chapter {chapter_num}")
        return []

    verses_data = []
    
    # We find all <sub> elements which represent verse numbers
    subs = content_div.find_all('sub', id=re.compile(r'^\d+$'))
    
    for sub in subs:
        verse_num_str = sub.get('id')
        if not verse_num_str.isdigit():
            continue
        verse_num = int(verse_num_str)
        
        # The text for this verse is everything after this <sub> tag until the next <sub> tag or end of paragraph
        verse_nodes = []
        current_node = sub.next_sibling
        while current_node:
            if current_node.name == 'sub' and current_node.get('id', '').isdigit():
                break  # Next verse starts
            
            if current_node.name not in ['p', 'div', 'br', 'hr']:
                if hasattr(current_node, 'get_text'):
                    verse_nodes.append(current_node.get_text(strip=False))
                elif isinstance(current_node, str):
                    verse_nodes.append(str(current_node))
            else:
                verse_nodes.append(" ")
            
            current_node = current_node.next_sibling
            
            if current_node is None:
                parent = sub.parent
                if parent:
                    next_p = parent.find_next_sibling(['p', 'div'])
                    if next_p:
                        first_sub = next_p.find('sub', id=re.compile(r'^\d+$'))
                        if first_sub:
                            for child in next_p.contents:
                                if child == first_sub:
                                    break
                                if hasattr(child, 'get_text'):
                                    verse_nodes.append(child.get_text(strip=False))
                                elif isinstance(child, str):
                                    verse_nodes.append(str(child))
                        else:
                            verse_nodes.append(next_p.get_text(separator=' ', strip=False))
        
        verse_text = ''.join(verse_nodes).strip()
        verse_text = re.sub(r'\s+', ' ', verse_text).strip()
        
        if verse_text:
            verses_data.append({
                'translation_id': TRANSLATION_ID,
                'book_id': BOOK_ID,
                'chapter': chapter_num,
                'verse_num': verse_num,
                'verse_text': verse_text
            })

    # Extract verse 1
    first_sub = content_div.find('sub', id='2')
    if first_sub and not any(v['verse_num'] == 1 for v in verses_data):
        verse1_nodes = []
        for p in content_div.find_all('p'):
            for child in p.contents:
                if child == first_sub:
                    break
                if child.name == 'sub' and child.get('id', '').isdigit():
                    continue
                if hasattr(child, 'get_text'):
                    verse1_nodes.append(child.get_text(strip=False))
                elif isinstance(child, str):
                    verse1_nodes.append(str(child))
            if first_sub in p.contents:
                break
        
        verse1_text = ''.join(verse1_nodes).strip()
        verse1_text = re.sub(r'\s+', ' ', verse1_text).strip()
        
        # Remove trailing verse numbers if accidentally included
        verse1_text = re.sub(r'\s*\d+$', '', verse1_text).strip()
        
        if verse1_text:
            verses_data.insert(0, {
                'translation_id': TRANSLATION_ID,
                'book_id': BOOK_ID,
                'chapter': chapter_num,
                'verse_num': 1,
                'verse_text': verse1_text
            })

    return verses_data

def main():
    print("Starting Matthew scraper...")
    all_verses = []
    for ch in range(1, 29):
        verses = scrape_chapter(ch)
        print(f"Chapter {ch}: found {len(verses)} verses")
        all_verses.extend(verses)
        
    print(f"Total verses scraped: {len(all_verses)}")
    
    # Insert to Supabase in batches
    batch_size = 100
    for i in range(0, len(all_verses), batch_size):
        batch = all_verses[i:i+batch_size]
        try:
            res = supabase.table('verses').upsert(batch, on_conflict='translation_id,book_id,chapter,verse_num').execute()
            print(f"Inserted batch {i//batch_size + 1}")
        except Exception as e:
            print(f"Error inserting batch {i//batch_size + 1}: {e}")

if __name__ == "__main__":
    main()

File: scripts/scrape_matthew.js
import { createClient } from '@supabase/supabase-js';
import * as cheerio from 'cheerio';
import dotenv from 'dotenv';
import fs from 'fs';
import { execSync } from 'child_process';

dotenv.config();

const SUPABASE_URL = process.env.VITE_SUPABASE_URL;
const SUPABASE_KEY = process.env.VITE_SUPABASE_SERVICE_ROLE_KEY || process.env.VITE_SUPABASE_ANON_KEY;
const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);

const TRANSLATION_ID = 1;
const BOOK_ID = 'mat';

function fetchHtml(url) {
    try {
        const stdout = execSync(`curl -sL "${url}"`);
        return stdout.toString();
    } catch (error) {
        throw new Error(`Failed to fetch ${url} using curl: ${error.message}`);
    }
}

async function scrapeChapter(chapterNum) {
    console.log(`Scraping chapter ${chapterNum}...`);
    const url = chapterNum === 1 
        ? "https://augustino.net/tin-mung-theo-thanh-mat-theu" 
        : `https://augustino.net/tin-mung-theo-thanh-mat-theu-chuong-${chapterNum}`;
    
    try {
        const html = await fetchHtml(url);
        const $ = cheerio.load(html);
        const contentDiv = $('#page-content').length ? $('#page-content') : $('.content');
        
        if (!contentDiv.length) {
            console.error(`Could not find content for chapter ${chapterNum}`);
            return [];
        }

        const versesData = [];
        const subs = contentDiv.find('sub');
        
        subs.each((i, el) => {
            const idAttr = $(el).attr('id');
            if (!idAttr || !/^\d+$/.test(idAttr)) return;
            
            const verseNum = parseInt(idAttr, 10);
            
            // Collect text until next <sub>
            let verseNodes = [];
            let current = el.nextSibling;
            
            while (current) {
                if (current.tagName === 'sub' && /^\d+$/.test($(current).attr('id'))) {
                    break;
                }
                
                if (!['p', 'div', 'br', 'hr'].includes(current.tagName)) {
                    if (current.nodeType === 3) { // Text node
                        verseNodes.push(current.data);
                    } else {
                        verseNodes.push($(current).text());
                    }
                } else {
                    verseNodes.push(" ");
                }
                
                current = current.nextSibling;
                
                if (!current) {
                    const parent = $(el).parent();
                    if (parent.length) {
                        const nextP = parent.next('p, div');
                        if (nextP.length) {
                            const firstSub = nextP.find('sub').filter((idx, s) => /^\d+$/.test($(s).attr('id'))).first();
                            if (firstSub.length) {
                                // Extract text up to firstSub
                                nextP.contents().each((idx, child) => {
                                    if (child === firstSub[0]) return false; // break loop
                                    if (child.nodeType === 3) {
                                        verseNodes.push(child.data);
                                    } else {
                                        verseNodes.push($(child).text());
                                    }
                                });
                            } else {
                                verseNodes.push(nextP.text());
                            }
                        }
                    }
                }
            }
            
            let verseText = verseNodes.join('').trim().replace(/\s+/g, ' ');
            
            if (verseText) {
                versesData.push({
                    translation_id: TRANSLATION_ID,
                    book_id: BOOK_ID,
                    chapter: chapterNum,
                    verse_num: verseNum,
                    verse_text: verseText
                });
            }
        });

        // Handle verse 1 implicitly (before sub id="2")
        const firstSub = contentDiv.find('sub#2');
        if (firstSub.length && !versesData.some(v => v.verse_num === 1)) {
            let verse1Nodes = [];
            let stop = false;
            contentDiv.find('p, div').each((i, el) => {
                if (stop) return false;
                $(el).contents().each((j, child) => {
                    if (child === firstSub[0]) {
                        stop = true;
                        return false;
                    }
                    if (child.tagName === 'sub' && /^\d+$/.test($(child).attr('id'))) {
                        return; // continue
                    }
                    if (child.nodeType === 3) {
                        verse1Nodes.push(child.data);
                    } else {
                        verse1Nodes.push($(child).text());
                    }
                });
            });
            let verse1Text = verse1Nodes.join('').trim().replace(/\s+/g, ' ');
            verse1Text = verse1Text.replace(/\s*\d+$/, '').trim(); // Remove trailing digits if accidentally included
            
            if (verse1Text) {
                versesData.unshift({
                    translation_id: TRANSLATION_ID,
                    book_id: BOOK_ID,
                    chapter: chapterNum,
                    verse_num: 1,
                    verse_text: verse1Text
                });
            }
        }

        return versesData;

    } catch (e) {
        console.error(`Error scraping chapter ${chapterNum}:`, e.message);
        return [];
    }
}

async function main() {
    console.log("Starting Matthew scraper using Node.js...");
    const allVerses = [];
    
    for (let ch = 1; ch <= 28; ch++) {
        const verses = await scrapeChapter(ch);
        console.log(`Chapter ${ch}: found ${verses.length} verses`);
        allVerses.push(...verses);
    }
    
    console.log(`Total verses scraped: ${allVerses.length}`);
    fs.writeFileSync('scripts/scraped_matthew.json', JSON.stringify(allVerses, null, 2));
    
    // Generate SQL file
    let sqlContent = `-- Dữ liệu Tin Mừng Mát-thêu (1071 câu)\n`;
    sqlContent += `INSERT INTO verses (translation_id, book_id, chapter, verse_num, verse_text) VALUES\n`;
    
    const values = allVerses.map(v => {
        const text = v.verse_text.replace(/'/g, "''");
        return `(${v.translation_id}, '${v.book_id}', ${v.chapter}, ${v.verse_num}, '${text}')`;
    });
    
    sqlContent += values.join(',\n') + `\nON CONFLICT (translation_id, book_id, chapter, verse_num) DO NOTHING;\n`;
    
    fs.writeFileSync('database/matthew_insert.sql', sqlContent);
    console.log("Generated database/matthew_insert.sql successfully!");
}

main();
